doc/designs/ech-api.md

```diff
@@ -0,0 +1,590 @@
+
+After a call to ``OSSL_ech_find_echconfigs()``, the application can make a
+sequence of calls to ``SSL_ech_set1_echconfig()`` for each of the ECHConfig
+values found.  (The various output buffers must be freed by the client
+afterwards, see the example code in
+[``test/ech_test.c``](https://github.com/sftcd/openssl/blob/ECH-draft-13c/test/ech_test.c).)
+
+Clients can additionally more directly control the values to be used for inner
+and outer SNI and ALPN values via specific APIs. This allows a client to
+override the ``public_name`` present in an ECHConfigList that will otherwise
+be used for the outer SNI. The ``no_outer`` input allows a client to emit an
+outer CH with no SNI at all.
+
+```c
+int SSL_ech_set_server_names(SSL *s, const char *inner_name,
+                             const char *outer_name, int no_outer);
+int SSL_ech_set_outer_server_name(SSL *s, const char *outer_name, int no_outer);
+int SSL_ech_set_outer_alpn_protos(SSL *s, const unsigned char *protos,
+                                  unsigned int protos_len);
+int SSL_CTX_ech_set_outer_alpn_protos(SSL_CTX *s, const unsigned char *protos,
+                                      unsigned int protos_len);
+```
+
+If a client attempts ECH but that fails, or sends an ECH-GREASEd CH, to
+an ECH-supporting server, then that server may return an ECH "retry-config"
+value that the client could choose to use in a subsequent connection. The
+client can detect this situation via the ``SSL_ech_get_status()`` API and
+can access the retry config value via:
+
+```c
+int SSL_ech_get_retry_config(SSL *s, unsigned char **ec, size_t *eclen);
+```
+
+Clients that need fine control over which ECHConfig (from those available) will
+be used, can query the SSL connection, retrieving information about the set of
+ECHConfig values available, and then, if desired, down-select to one of those,
+e.g., based on the ``public_name`` that will be used. This would enable a
+client that selects the server address to use based on IP address hints that
+can also be present in an HTTPS/SCVB resource record to ensure that the correct
+matching ECHConfig is used. The information is presented to the client using
+the ``OSSL_ECH_INFO`` type, which provides a simplified view of ECHConfig data,
+but where each element of an array corresponds to exactly one ECH public value
+and set of names.
+
+```c
+/*
+ * Application-visible form of ECH information from the DNS, from config
+ * files, or from earlier API calls. APIs produce/process an array of these.
+ */
+typedef struct ossl_ech_info_st {
+    int index; /* externally re-usable reference to this value */
+    char *public_name; /* public_name from API or ECHConfig */
+    char *inner_name; /* server-name (for inner CH if doing ECH) */
+    char *outer_alpns; /* outer ALPN string */
+    char *inner_alpns; /* inner ALPN string */
+    char *echconfig; /* a JSON-like version of the associated ECHConfig */
+} OSSL_ECH_INFO;
```

Comment:
Why is this a structure and not an opaque object?

Reply:
That's intended for the unusual "advanced" uses where a client might want to compare against other SvcParamKey values from an HTTPS/SVCB RR. That kind of API might be needed e.g. if https://datatracker.ietf.org/doc/draft-pauly-v6ops-happy-eyeballs-v3/ turns into a useful thing.

FWIW, I do agree however that kind of functionality could be omitted from a 1st release of ECH in the library. I included it because I have it and something of that nature will be needed in the not too distant, but it could be the subject of an indpendent PR after the ECH feature is merged to master. OTOH, I think it may be better to keep it as part of the feature branch in the expectation that by the time the ECH code is ready to merge with master, the "advanced" client uses will be clear.